### ISSUE TITLE
Upgrading RazorGenerator.MSBuild from 2.2.6 to 2.3.4 breaks rendering

### ISSUE THREAD

**Issue by asbjornu** [RazorGenerator/RazorGenerator#29]
I just tried to upgrade RazorGenerator.MsBuild from version `2.2.6` to `2.3.4` and RazorGenerator.Mvc from version `2.2.3` to version `2.3.5`. For reasons unbeknownst to me, that broke rendering of all views. In some circumstances, I get this exception:
```
The view 'Index' or its master was not found or no view engine supports the searched locations. The following locations were searched: ~/Areas/Redacted/Views/Projects/Index.cshtml ~/Areas/Redacted/Views/Shared/Index.cshtml ~/Views/Projects/Index.cshtml ~/Views/Shared/Index.cshtml
```
With the following stack trace:
```c#
at System.Web.Mvc.ViewResult.FindView(ControllerContext context) at System.Web.Mvc.ViewResultBase.ExecuteResult(ControllerContext context) at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultWithFilters(ControllerContext controllerContext, IList`1 filters, ActionResult actionResult) at System.Web.Mvc.Async.AsyncControllerActionInvoker.<>c__DisplayClass21.b__1e(IAsyncResult asyncResult)
```
In other circumstances, I get no exception, but a view rendered without a layout. This brings up a lot of questions I would love to have answered:

1. Has there been any breaking changes between the mentioned versions of these two packages?
2. How does the two packages relate to each other?
3. How do the two packages maintain backward compatibility?
4. How is the backward compatibility between the two packages expressed in terms of the version number? Can we expect every minor version of `2` to be compatible both ways?
3. Why is the last version of RazorGenerator.MsBuild `2.3.4` while it's `2.3.5` for RazorGenerator.Mvc?

Thanks!

**Comment by davidebbo** [comment]
Note sure if `2.3.4` vs `2.3.5` is the issue, but just in case I pushed `RazorGenerator.MsBuild 2.3.5`. Can you see if that makes a difference?

If it's still broken, I'll let @pranavkm comment as he's the RazorGenerator.MsBuild expert.

**Comment by pranavkm** [comment]
We typically don't make breaking changes between point releases, but there were a few changes to how the msbuild target locates the cshtml files from your project. We could start by verifying the views are actually being compiled in to your assembly.

Could you check if the generated files appear in the `obj` directory of your project and that the views get compiled in to your project assembly?

**Comment by asbjornu** [comment]
@davidebbo Upgrading to `2.3.5` had no effect, sadly. @pranavkm Yes, the folder `obj/CodeGen/Views` contain all of the views. In dotPeek I can see that the compiled assembly contains an `ASP` namespace in which classes like `_Views_Compare_Index_cshtml` reside. Does this sound correct? What else might cause the views from being loaded correctly?

I'm using the `CompositePrecompiledMvcEngine`, if that makes any difference.

**Comment by pranavkm** [comment]
@asbjornu that looks right so far. Can you check if the `VirtualPathAttribute` for the precompiled file in `obj/CodeGen/Areas/Redacted/Views/Index.cs` looks correct (i.e. it maps to `~/Areas/Redacted/Views/Shared/Index.cshtml`)? You could also try debugging `CompositePrecompiledMvcEngine` - dropping the cs file from `https://github.com/RazorGenerator/RazorGenerator/blob/master/RazorGenerator.Mvc/CompositePrecompiledMvcEngine.cs` into your application should suffice.

**Comment by asbjornu** [comment]
@pranavkm Thanks for pushing me in the right direction! After comparing the contents of the `CodeGen` folder of the branch where RazorGenerator is upgraded and where it's still at version `2.2.6` (master), I discovered that the upgraded branch was missing a lot of `.cshtml.cs` files. For some reason, the missing views were included in the `.csproj` file with `<None>`. Changing them to `<Content>` fixed the problem.

Investigating this further, there's actually three views that have a `None` build action in master:
* `~/Views/Projects/Index.cshtml`
* `~/Views/Shared/_Layout.cshtml`
* `~/Views/Shared/_Edit.cshtml`

On the master branch, running on RazorGenerator.MsBuild version `2.2.6` and RazorGenerator.Mvc version `2.2.3`, this works just fine. After upgrading to version `2.3.5`, having `Index.cshtml` as `None` breaks. I assume shared views don't need to be `Content` as they are implicitly compiled as part of all other views?

**Comment by pranavkm** [comment]
@asbjornu seems like it's a behavioral change introduced as part of #8 (https://github.com/RazorGenerator/RazorGenerator/pull/8. We're only pulling Content files from Views:

https://github.com/RazorGenerator/RazorGenerator/blob/master/RazorGenerator.MsBuild/RazorGenerator.MsBuild.targets#L14-L15

My guess is we could additionally include all files marked with `None` in addition to the ones with `Content` by default. @cjberg, thoughts?

**Comment by asbjornu** [comment]
@pranavkm I see. If you don't include `None` files, perhaps you could issue a warning if a `.cshtml` file is included with `None`?

**Comment by cjberg** [comment]
@pranavkm You're correct, `None` should be included as it can be used when one e.g. don't want to include the cshtml file for deployment. (I didn't have a single view file in that item group when I ran tests, because my main project needs to deploy the cshtml files and fails the build if one isn't in the `Content` group. That's so typical!)

I think we only need to add `;@(None)` in the targets file, like this:
```
<RazorSrcFiles Condition=" '@(RazorSrcFiles)' == '' and '%(Extension)' == '.cshtml' "
    Include="@(Content);@(None)" />
```
I'm on vacation at the moment, so I can't test it or create a pull request, but I see no reason why it wouldn't work.

**Comment by asbjornu** [comment]
@cjberg Good point. Skip the warning. :)

**Comment by pranavkm** [comment]
Fixed in 2011802019ba2ad5f73208d1c50e5dd009b839bd. Package uploaded to NuGet - https://www.nuget.org/packages/RazorGenerator.MsBuild

**Comment by asbjornu** [comment]
Brilliant! :)
